plugins/builds.js

```diff
@@ -0,0 +1,131 @@
+const DEFAULT_BYTES = 1024 * 1024 * 1024; // 1GB
+
+/**
+ * Builds Plugin
+ * @method register
+ * @param  {Hapi}     server                Hapi Server
+ * @param  {Object}   options               Configuration
+ * @param  {Integer}  options.expiresIn     How long to keep it around
+ * @param  {Integer}  options.maxByteSize   Maximum Bytes to accept
+ * @param  {Function} next                  Function to call when done
+ */
+exports.register = (server, options, next) => {
+    const cache = server.cache({
+        segment: 'builds',
+        expiresIn: options.expiresIn || DEFAULT_TTL
+    });
+
+    server.expose('stats', cache.stats);
+    server.route([{
+        method: 'GET',
+        path: '/builds/{id}/{artifact*}',
+        config: {
+            description: 'Read build artifacts',
+            notes: 'Get an artifact from a specific build',
+            tags: ['api', 'builds'],
+            handler: (request, reply) => {
+                const buildId = request.params.id;
+                const artifact = request.params.artifact;
+                const id = `${buildId}-${artifact}`;
+
+                cache.get(id, (err, value) => {
+                    if (err) {
+                        return reply(err);
+                    }
+                    if (err || value === null) {
+                        return reply(boom.notFound());
+                    }
+
+                    // @TODO put cache headers in here
+                    const response = reply(Buffer.from(value.c.data));
+
+                    response.headers = value.h;
+
+                    return response;
+                });
+            },
+            validate: {
+                params: {
+                    id: SCHEMA_BUILD_ID,
+                    artifact: SCHEMA_ARTIFACT_ID
+                }
+            }
+        }
+    }, {
+        method: 'PUT',
+        path: '/builds/{id}/{artifact*}',
+        config: {
+            description: 'Write build artifacts',
+            notes: 'Write an artifact from a specific build',
+            tags: ['api', 'builds'],
+            payload: {
+                maxBytes: options.maxByteSize || DEFAULT_BYTES,
+                parse: false
+            },
+            auth: {
+                strategies: ['token'],
+                scope: ['build']
+            },
+            plugins: {
+                'hapi-swagger': {
+                    security: [{ token: [] }]
+                }
+            },
+            handler: (request, reply) => {
+                const username = request.auth.credentials.username;
+                const buildId = request.params.id;
+                const artifact = request.params.artifact;
+                const id = `${buildId}-${artifact}`;
+                const contents = {
+                    c: request.payload,
+                    h: {}
+                };
+                const size = Buffer.byteLength(request.payload);
+
+                if (username !== buildId) {
```

Comment:
I'm confused about this. In the `GET`, the `id` is `<build_id>-<artifact>`. When will it be equal to the username?